File: zeroday-server/src/main/kotlin/com/zeroday/model/Zone.kt
package com.zeroday.model

import kotlinx.serialization.Serializable
import java.util.UUID

enum class ZoneFaction(val displayName: String) {
    Neutral("Neutral"),
    Syndicate("Syndicate"),
    CorpNet("CorpNet"),
    GhostNet("GhostNet"),
    FreeWorld("FreeWorld"),
    ZeroDay("ZeroDay"),
    LawEnforcement("Law Enforcement")
}

enum class ZoneState { Safe, Contested, Controlled, Warzone, Locked }

@Serializable
data class Zone(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val description: String,
    var controllingFaction: ZoneFaction = ZoneFaction.Neutral,
    var state: ZoneState = ZoneState.Safe,
    var controlLevel: Int = 50,
    val maxControlLevel: Int = 100,
    var securityLevel: Int = 1,
    val requiredLevel: Int = 1,
    val connectedZoneIds: List<String> = emptyList(),
    var hasBoss: Boolean = false,
    var bossName: String = "",
    var bossLevel: Int = 1,
    var threatLevel: Int = 0
)

@Serializable
data class ZoneSnapshot(
    val id: String,
    val name: String,
    val description: String,
    val controllingFaction: String,
    val state: String,
    val controlLevel: Int,
    val maxControlLevel: Int,
    val securityLevel: Int,
    val requiredLevel: Int,
    val connectedZoneIds: List<String>,
    val hasBoss: Boolean,
    val bossName: String,
    val bossLevel: Int,
    val threatLevel: Int
) {
    companion object {
        fun from(zone: Zone): ZoneSnapshot = ZoneSnapshot(
            id = zone.id,
            name = zone.name,
            description = zone.description,
            controllingFaction = zone.controllingFaction.name,
            state = zone.state.name,
            controlLevel = zone.controlLevel,
            maxControlLevel = zone.maxControlLevel,
            securityLevel = zone.securityLevel,
            requiredLevel = zone.requiredLevel,
            connectedZoneIds = zone.connectedZoneIds,
            hasBoss = zone.hasBoss,
            bossName = zone.bossName,
            bossLevel = zone.bossLevel,
            threatLevel = zone.threatLevel
        )
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Shared/Models/Zone.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ZeroDayMMO.Shared.Models;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ZoneFaction
{
    [JsonPropertyName("neutral")]
    Neutral,
    [JsonPropertyName("syndicate")]
    Syndicate,
    [JsonPropertyName("corpnet")]
    CorpNet,
    [JsonPropertyName("ghostnet")]
    GhostNet,
    [JsonPropertyName("freeworld")]
    FreeWorld,
    [JsonPropertyName("zeroday")]
    ZeroDay,
    [JsonPropertyName("law_enforcement")]
    LawEnforcement
}

public static class ZoneFactionExtensions
{
    public static string DisplayName(this ZoneFaction faction) => faction switch
    {
        ZoneFaction.Neutral => "Neutral",
        ZoneFaction.Syndicate => "Syndicate",
        ZoneFaction.CorpNet => "CorpNet",
        ZoneFaction.GhostNet => "GhostNet",
        ZoneFaction.FreeWorld => "FreeWorld",
        ZoneFaction.ZeroDay => "ZeroDay",
        ZoneFaction.LawEnforcement => "Law Enforcement",
        _ => "Neutral"
    };
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ZoneState
{
    [JsonPropertyName("safe")]
    Safe,
    [JsonPropertyName("contested")]
    Contested,
    [JsonPropertyName("controlled")]
    Controlled,
    [JsonPropertyName("warzone")]
    Warzone,
    [JsonPropertyName("locked")]
    Locked
}

public class Zone
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = Guid.NewGuid().ToString();

    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("description")]
    public string Description { get; set; } = "";

    [JsonPropertyName("controlling_faction")]
    public ZoneFaction ControllingFaction { get; set; } = ZoneFaction.Neutral;

    [JsonPropertyName("state")]
    public ZoneState State { get; set; } = ZoneState.Safe;

    [JsonPropertyName("control_level")]
    public int ControlLevel { get; set; } = 50;

    [JsonPropertyName("max_control_level")]
    public int MaxControlLevel { get; set; } = 100;

    [JsonPropertyName("security_level")]
    public int SecurityLevel { get; set; } = 1;

    [JsonPropertyName("required_level")]
    public int RequiredLevel { get; set; } = 1;

    [JsonPropertyName("connected_zone_ids")]
    public List<string> ConnectedZoneIds { get; set; } = new();

    [JsonPropertyName("has_boss")]
    public bool HasBoss { get; set; } = false;

    [JsonPropertyName("boss_name")]
    public string BossName { get; set; } = "";

    [JsonPropertyName("boss_level")]
    public int BossLevel { get; set; } = 1;

    [JsonPropertyName("threat_level")]
    public int ThreatLevel { get; set; } = 0;
}

public class ZoneSnapshot
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = "";

    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("description")]
    public string Description { get; set; } = "";

    [JsonPropertyName("controlling_faction")]
    public string ControllingFaction { get; set; } = "";

    [JsonPropertyName("state")]
    public string State { get; set; } = "";

    [JsonPropertyName("control_level")]
    public int ControlLevel { get; set; }

    [JsonPropertyName("max_control_level")]
    public int MaxControlLevel { get; set; }

    [JsonPropertyName("security_level")]
    public int SecurityLevel { get; set; }

    [JsonPropertyName("required_level")]
    public int RequiredLevel { get; set; }

    [JsonPropertyName("connected_zone_ids")]
    public List<string> ConnectedZoneIds { get; set; } = new();

    [JsonPropertyName("has_boss")]
    public bool HasBoss { get; set; }

    [JsonPropertyName("boss_name")]
    public string BossName { get; set; } = "";

    [JsonPropertyName("boss_level")]
    public int BossLevel { get; set; }

    [JsonPropertyName("threat_level")]
    public int ThreatLevel { get; set; }

    public static ZoneSnapshot From(Zone zone) => new()
    {
        Id = zone.Id,
        Name = zone.Name,
        Description = zone.Description,
        ControllingFaction = zone.ControllingFaction.ToString(),
        State = zone.State.ToString(),
        ControlLevel = zone.ControlLevel,
        MaxControlLevel = zone.MaxControlLevel,
        SecurityLevel = zone.SecurityLevel,
        RequiredLevel = zone.RequiredLevel,
        ConnectedZoneIds = new List<string>(zone.ConnectedZoneIds),
        HasBoss = zone.HasBoss,
        BossName = zone.BossName,
        BossLevel = zone.BossLevel,
        ThreatLevel = zone.ThreatLevel
    };
}
